### PR TITLE
Nether Portal Jumping & Creation Code

### PR DESCRIPTION
**Pull request by lkolbly**
Per #1856, this code implements a portal scanner which finds and creates nether portals when entities move between the overworld and the nether.

It also changes https://github.com/mc-server/MCServer/commit/dae9e579#diff-97bcf8a96a22aec81982926aa1462363R298 because StringToInteger returns true if the conversion was successful. I think.

### CODE REVIEW

File: src/BlockID.cpp [hunk 1]
```diff
@@ -295,7 +295,7 @@ eDimension StringToDimension(const AString & a_DimensionString)
 {
 	// First try decoding as a number
 	int res;
-	if (!StringToInteger(a_DimensionString, res))
```
Comment (worktycho) [inline src/BlockID.cpp]: Thanks, I got a few of these wrong.

File: src/Chunk.cpp [hunk 2]
```diff
@@ -640,6 +640,7 @@ void cChunk::Tick(std::chrono::milliseconds a_Dt)
 		else if ((*itr)->IsWorldTravellingFrom(m_World))
 		{
 			// Remove all entities that are travelling to another world
+			LOGD("Removing entity from %d,%d that's travelling between worlds.", m_PosX, m_PosZ);
```
Comment (madmaxoft) [inline src/Chunk.cpp]: Usually we log chunk coords enclosed in `[]` and block coords enclosed in `{}`. Also, add a space after the comma.

File: src/Entities/Entity.cpp [hunk 3]
```diff
@@ -1259,10 +1259,349 @@ void cEntity::DetectCacti(void)
...
+	}
+	else
+	{
+		m_PortalLoc.x += AcrossOffset;
+		m_PortalLoc.z += OutOffset;
+	}
+
+	m_Entity->ScheduleMoveToWorld(m_World, m_PortalLoc);
+	delete this;
+}
```
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: This entire class should be moved into a separate file (Entity.cpp is already huge without it). Also, separate each function by 5 empty lines.

Comment (lkolbly) [inline src/Entities/Entity.cpp]: Alrighty. I noticed you guys have a vcproj file in src/ - I don't have Visual Studio, do I need to do anything to update that file or can I just create the file and add it to CMakeLists? (and the AllFiles.lst)

**Comment by madmaxoft** [comment]
Please try to use the same coding style. We use a space after all commas, and a space around pointer and reference declarations (`cWorld * a_World` etc.) There's too many of these to report each one. Unfortunately the basic stylechecking script cannot detect these reliably, so you'll need to do it by hand.

**Comment by lkolbly** [comment]
Okay, I'll go back through as I move it to a new file.

File: src/Entities/Entity.cpp (hunk 3 above)
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: There's no vcproj file, there's only a .vcproj.user file which is a template for per-user settings in VS. You only need to add the file to CMakeLists (it will be added to AllFiles.lst automatically)

File: src/Entities/Entity.cpp [hunk 4]
```diff
@@ -1259,10 +1259,349 @@ void cEntity::DetectCacti(void)
...
+	return true;
+}
+
+
+
+bool cNetherPortalScanner::OnAllChunksAvailable(void)
+{
+	if (m_FoundPortal)
+	{
+		m_PortalLoc = m_PortalLoc;
```
Comment (NiLSPACE) [inline src/Entities/Entity.cpp]: I doubt this is needed.

Comment (lkolbly) [inline src/Entities/Entity.cpp]: Good catch, thanks.

File: src/BlockID.cpp (hunk 1 above)
Comment (worktycho) [inline src/BlockID.cpp]: I'm going to manually apply this patch to master, because its an important fix.

**Comment by lkolbly** [comment]
Okay, I think I got it compliant with the coding convention in bad47a8, except maybe local variable names, I don't know what the convention is for those. I didn't see any commas without spaces, though, so I'm probably looking at the wrong commas?

File: src/Entities/Entity.cpp [hunk 5]
```diff
@@ -1260,9 +1261,25 @@ void cEntity::DetectCacti(void)
...
+
 
 bool cEntity::DetectPortal()
 {
+	if (m_ReadyToChangeWorlds)
+	{
+		m_ReadyToChangeWorlds = false;
+		MoveToWorld(m_NewWorld, false, m_NewWorldPosition);
+		return true;
+	}
```
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: This block deserves a comment above it explaining what it is doing.

File: src/Entities/Entity.cpp [hunk 6]
```diff
@@ -1312,8 +1329,15 @@ bool cEntity::DetectPortal()
...
 						((cPlayer *)this)->GetClientHandle()->SendRespawn(dimOverworld);
 					}
-					
-					return MoveToWorld(cRoot::Get()->CreateAndInitializeWorld(GetWorld()->GetLinkedOverworldName()), false);
+
+					Vector3d TargetPos = GetPosition();
+					TargetPos.x *= 8.0;
+					TargetPos.z *= 8.0;
+
+					cWorld *TargetWorld = cRoot::Get()->CreateAndInitializeWorld(GetWorld()->GetLinkedOverworldName(), dimNether, GetWorld()->GetName());
```
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: space after *

File: src/Entities/Entity.cpp [hunk 7]
```diff
@@ -1329,8 +1353,15 @@ bool cEntity::DetectPortal()
...
 						((cPlayer *)this)->GetClientHandle()->SendRespawn(dimNether);
 					}
-					
-					return MoveToWorld(cRoot::Get()->CreateAndInitializeWorld(GetWorld()->GetLinkedNetherWorldName(), dimNether, GetWorld()->GetName()), false);
+
+					Vector3d TargetPos = GetPosition();
+					TargetPos.x /= 8.0;
+					TargetPos.z /= 8.0;
+
+					cWorld *TargetWorld = cRoot::Get()->CreateAndInitializeWorld(GetWorld()->GetLinkedNetherWorldName(), dimNether, GetWorld()->GetName());
```
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: space after *

File: src/Entities/Entity.h [hunk 8]
```diff
@@ -392,15 +392,19 @@ class cEntity
 	/// Teleports to the coordinates specified
 	virtual void TeleportToCoords(double a_PosX, double a_PosY, double a_PosZ);
 
+	void ScheduleMoveToWorld(cWorld *a_World, Vector3d a_NewPosition);
```
Comment (madmaxoft) [inline src/Entities/Entity.h]: space after *

File: src/Entities/Entity.h [hunk 9]
```diff
@@ -530,6 +534,10 @@ class cEntity
 	eEntityType m_EntityType;
 	
 	cWorld * m_World;
+
+	bool m_ReadyToChangeWorlds = false;
+	cWorld *m_NewWorld;
```
Comment (madmaxoft) [inline src/Entities/Entity.h]: space after *

File: src/NetherPortalScanner.cpp [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+	m_BuildPlatform = true;
+	m_Dir = 0;
+	m_Position = a_DestPosition;
+	m_PortalLoc = a_DestPosition;//TODO: Cast!
+	m_MaxY = a_MaxY;
+
+	int MinX = (int)floor((m_Position.x - SearchRadius) / cChunkDef::Width);
+	int MinZ = (int)floor((m_Position.z - SearchRadius) / cChunkDef::Width);
+	int MaxX = (int)ceil((m_Position.x + SearchRadius) / cChunkDef::Width);
+	int MaxZ = (int)ceil((m_Position.z + SearchRadius) / cChunkDef::Width);
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: Use `FloorC` and `CeilC` instead - they do the static cast to `int` for you.

File: src/NetherPortalScanner.cpp [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+			m_Dir = 1;
+		}
+	}
+	else
+	{
+		// Scan the area for a suitable location
+		int minx = (int)floor(m_Position.x) - BuildSearchRadius;
+		int minz = (int)floor(m_Position.z) - BuildSearchRadius;
+		int maxx = (int)floor(m_Position.x) + BuildSearchRadius;
+		int maxz = (int)floor(m_Position.z) + BuildSearchRadius;
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: `FloorC`

File: src/NetherPortalScanner.cpp [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+
+
+void cNetherPortalScanner::OnDisabled(void)
+{
+	// Now we actually move the player
+	if (!m_FoundPortal)
+	{
+		// Build a new nether portal.
+		LOGD("Building nether portal at {%d,%d,%d}", m_PortalLoc.x, m_PortalLoc.y, m_PortalLoc.z);
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: Spaces after commas

File: src/NetherPortalScanner.h [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+
+// This is the chunk stay which finds nearby nether portals
+class cNetherPortalScanner : public cChunkStay
+{
+public:
+	cNetherPortalScanner(cEntity * a_MovingEntity, cWorld * a_DestinationWorld, Vector3d a_DestPosition, int a_MaxY);
+	void OnChunkAvailable(int a_ChunkX, int a_ChunkY);
+	bool OnAllChunksAvailable(void);
+	void OnDisabled(void);
```
Comment (madmaxoft) [inline src/NetherPortalScanner.h]: These functions are `virtual` and should have an `override` specifier.

File: src/NetherPortalScanner.cpp [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+		}
+	}
+	return true;
+}
+
+
+
+
+
+void cNetherPortalScanner::BuildNetherPortal(Vector3i a_Location, int a_Direction, int a_IncludePlatform)
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: `a_IncludePlatform` should be a `bool`

File: src/NetherPortalScanner.h [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	/** The world we're moving the entity to. */
+	cWorld * m_World;
+
+	/** Whether we found a portal during the loading of the chunks. */
+	bool m_FoundPortal;
+
+	/** Whether to build a platform. True if we couldn't build the portal on solid ground */
+	bool m_BuildPlatform;
+
+	/** The direction of the portal. 0=long axis along X, 1=long axis along Z */
```
Comment (madmaxoft) [inline src/NetherPortalScanner.h]: Spaces around `=`

File: src/NetherPortalScanner.cpp [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+		}
+
+		if (Possibilities.size() > 0)
+		{
+			m_BuildPlatform = false;
+
+			// Find the nearest
+			double DistanceToClosest = (Possibilities[0] - m_Position).SqrLength();
+			Vector3i Closest = Possibilities[0];
+			for (std::vector<Vector3i>::iterator itr = Possibilities.begin(); itr != Possibilities.end(); itr++)
```
Comment (tigerw) [inline src/NetherPortalScanner.cpp]: `for (const auto & itr : Possibilities)`?

**Comment by lkolbly** [comment]
What's the preferred way to convert between Vector types, e.g. from Vector3i to a Vector3d?

File: src/NetherPortalScanner.cpp (hunk 16 above)
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: Not an `itr` :P

**Comment by madmaxoft** [comment]
I don't think there's any preference. There's a simple copy-constructor provided within the vector template, and for converting to a `Vector3i` there's also the `Floor()` function defined on the class (`Ceil()` is missing right now, but could be added with minimal effort).

**Comment by lkolbly** [comment]
Okay. I think I've fixed the convention errors.

File: src/Entities/Entity.cpp (hunk 5 above)
Comment (madmaxoft) [inline src/Entities/Entity.cpp]: I think a simple `/* If a move to another world was scheduled via ScheduleMoveToWorld(), move the entity now. */` would have been enough. This is too wordy and still doesn't explain much. Also note that comments inside functions shouldn't use the doxy-comment format (`/**`), only comments above declarations should use it.

Also the `m_ReadyToChangeWorlds` is a somewhat misleading name, `m_IsWorldChangeScheduled` might have been better.

File: src/NetherPortalScanner.cpp [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+cNetherPortalScanner::cNetherPortalScanner(cEntity * a_MovingEntity, cWorld * a_DestinationWorld, Vector3d a_DestPosition, int a_MaxY)
+{
+	m_Entity = a_MovingEntity;
+	m_World = a_DestinationWorld;
+	m_FoundPortal = false;
+	m_BuildPlatform = true;
+	m_Dir = 0;
+	m_Position = a_DestPosition;
+	m_PortalLoc = a_DestPosition.Floor();
+	m_MaxY = a_MaxY;
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: These should go in the initialization section of the constructor

File: src/NetherPortalScanner.cpp [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+	int z = a_Location.z;
+
+	// Clear a 3x4x4 area starting right above the base
+	for (int i = 0; i < SearchSolidBaseWidth; i++)
+	{
+		for (int j = 0; j < PortalLength; j++)
+		{
+			for (int k = 1; k < PortalHeight; k++)
+			{
+				if (a_Direction)
```
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: Don't use an `int` like a `bool`. Ints should be compared to explicit values, bools should be used directly.
`if (intValue == 4)` vs `if (boolValue)`

**Comment by lkolbly** [comment]
Alright, I just refactored m_Dir to use a Direction type which is either X or Y (I considered NORTH_SOUTH and EAST_WEST, but that didn't seem to map to the code as well).

**Comment by madmaxoft** [comment]
Is there any chance you could get this rebased onto current master and squashed into a single commit?

**Comment by lkolbly** [comment]
Sure thing. Done.

**Comment by madmaxoft** [comment]
I think this is merge-worthy. Opinions?

File: src/Entities/Entity.h [hunk 19]
```diff
@@ -530,6 +535,11 @@ class cEntity
 	eEntityType m_EntityType;
 	
 	cWorld * m_World;
+
+	/// State variables for ScheduleMoveToWorld.
+	bool m_IsWorldChangeScheduled = false;
```
Comment (NiLSPACE) [inline src/Entities/Entity.h]: Shouldn't this be initialized in the construcor? That is where we usualy initialize member variables right?

**Comment by NiLSPACE** [comment]
Apart from the question I have about the `m_IsWorldChangeScheduled` initialization it looks really good.

File: src/Entities/Entity.h (hunk 19 above)
Comment (tigerw) [inline src/Entities/Entity.h]: And also `/** */` for comment style.

**Comment by Seadragon91** [comment]
Please add yourself to the contributors file.

File: src/NetherPortalScanner.h [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+// This is the chunk stay which finds nearby nether portals
+class cNetherPortalScanner : public cChunkStay
+{
+public:
+	cNetherPortalScanner(cEntity * a_MovingEntity, cWorld * a_DestinationWorld, Vector3d a_DestPosition, int a_MaxY);
+	virtual void OnChunkAvailable(int a_ChunkX, int a_ChunkY) override;
+	virtual bool OnAllChunksAvailable(void) override;
+	virtual void OnDisabled(void) override;
+
+	enum Direction
```
Comment (tigerw) [inline src/NetherPortalScanner.h]: Maybe `enum class Direction`?

File: src/Entities/Entity.h (hunk 19 above)
Comment (lkolbly) [inline src/Entities/Entity.h]: Yes, it should be initialized in the constructor. For the comment, there was a previous block of several comments there that were triple slashes, so I changed the whole block in the commit I just now pushed.

**Comment by SphinxC0re** [comment]
Looks like it's ready for merge

File: src/Entities/Entity.h (hunk 19 above)
Comment (madmaxoft) [inline src/Entities/Entity.h]: Yeah, sorry for the comment style, we couldn't decide on the format beforehands, so one half of the code uses `///` and the other half uses `/** */`. We've since then decided to use the latter, but didn't fix the former yet. Good call about replacing them all.

**Comment by NiLSPACE** [comment]
After almost a month I finally updated my MCServer and tried this, and I noticed that when teleporting to an portal I get teleported 1 block next to it. That probably isn't intended right? You're supposed to teleport in another portal.

**Comment by SphinxC0re** [comment]
Also, you get teleported instantly but I think this is another issue

**Comment by NiLSPACE** [comment]
That's only in creative gamemode right? Because in vanilla you also get teleported instantly in creative mode.

File: src/NetherPortalScanner.cpp [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+
+
+
+void cNetherPortalScanner::OnChunkAvailable(int a_ChunkX, int a_ChunkZ)
+{
+	cChunkDef::BlockTypes blocks;
+	m_World->GetChunkBlockTypes(a_ChunkX, a_ChunkZ, blocks);
+
+	// Iterate through all of the blocks in the chunk
+	for (unsigned int i = 0; i < cChunkDef::NumBlocks; i++)
```
Comment (NiLSPACE) [inline src/NetherPortalScanner.cpp]: Looking at this again, shouldn't this be `i += 2` as suggested by @madmaxoft [here](https://github.com/cuberite/cuberite/issues/1856#issuecomment-105779965)? It would speed up the search for a nether portal by quite allot since it wouldn't have to check every block in the chunk.

**Comment by Seadragon91** [comment]
I tested this, I can not get back to the world from the nether trough the portal. Used the FAQ for creating a nether world.

**Comment by NiLSPACE** [comment]
That's probably because you haven't configured the world linking in the world.ini. The FAQ only describes how you can generate such a world.

**Comment by SphinxC0re** [comment]
Yes, this is a problem. You *have* to stand right in the portal to teleport. in vanilla you only need partial contact with the portal.

**Comment by tigerw** [comment]
> Yes, this is a problem. You have to stand right in the portal to teleport. in vanilla you only need partial contact with the portal.

This is to do with lack of bounding box detection in MCS at the present.

File: src/NetherPortalScanner.cpp (hunk 21 above)
Comment (lkolbly) [inline src/NetherPortalScanner.cpp]: Because of the way the chunks are stored (XZY?), doing i+2 here would result in a pattern that skips every other row. Note that it's iterating from 0 to NumBlocks, not 0 to 16 each axis - this results in a speedup in not having to call CoordinateToIndex.

**Comment by lkolbly** [comment]
@NiLSPACE 
> ... and I noticed that when teleporting to an portal I get teleported 1 block next to it.
That was intended, but can be changed easily.

**Comment by NiLSPACE** [comment]
I think it should be changed, because I got teleported in the portal frame once. It **could** also be possible to be teleported inside a wall if the portal is next to it.

**Comment by lkolbly** [comment]
I'm not sure you can since it places the portal in a 3x4x5 empty space, so there's always air on either side of the portal. But I can change it anyway.

**Comment by NiLSPACE** [comment]
But if I build a portal against a wall then there isn't an empty space around it. ^^

**Comment by lkolbly** [comment]
Oh, that's true. I hadn't thought about manually created portals. I will change that & the Y limit and make a PR after #2237 is closed. Having multiple forks is so much effort.

File: src/NetherPortalScanner.cpp (hunk 21 above)
Comment (madmaxoft) [inline src/NetherPortalScanner.cpp]: It would have been possible to have two loops, the inner one iterating over the 256 blocks in a single Y slice, and the outer one iterating over Y slices, skipping 2 slices out of each 3:
```cpp
for (unsigned y = 0; y < cChunkDef::Height; y += 3)
{
  unsigned BaseY = cChunkDef::Width * cChunkDef::Width * y;
  for (unsigned i = 0; i < 256; i++)
  {
    // Check block at index [BaseY + i]
  }
}
```
